product_multi_company/models/product_product.py

```diff
@@ -0,0 +1,47 @@
+# Copyright 2020 Kevin Graveman <k.graveman@onestein.nl>
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = ["multi.company.abstract", "product.product"]
+    _name = "product.product"
+    _description = "Product Variant (Multi-Company)"
+
+    company_ids = fields.Many2many(
+        string="Companies",
+        comodel_name="res.company",
+        default=lambda self: self._default_company_ids(),
```

Comment:
This default is going to act on module update, so you should add a pre-migration for adding manually the SQL column.

Reply:
As far as I've observed in the past, if a table does not yet exist and the field definition has a default or a compute method, then the table is created and populated accordingly on an upgrade. While testing this is also what happens, the `product_product_res_company_rel` table is created, but as self.env.company.ids is not relevant, it stays empty. This is what we want since we want the post-migration script to populate it.
I've tested it with products with and without company_ids set, and the results are as expected.